vllm_gaudi/distributed/kv_transfer/kv_connector/v1/hpu_nixl_connector.py

```diff
@@ -50,6 +53,43 @@ def _hpu_data_ptr(tensor_self):
     return original_data_ptr(tensor_self)
 
 
+def wait_for_save(self):
+    assert self.connector_worker is not None
+    assert isinstance(self._connector_metadata, NixlConnectorMetadata)
+    self.connector_worker.rewrite_kv_based_on_transfer_layout(self._connector_metadata)
+    if self.connector_worker.use_host_buffer and \
+       self.connector_worker.copy_blocks:
+        self.connector_worker.save_kv_to_host(self._connector_metadata)
+
+
+def rewrite_kv_based_on_transfer_layout(self, metadata: NixlConnectorMetadata):
+    decoder_tp_ratio = int(os.getenv('DECODER_TP_RATIO', 1))
+    if decoder_tp_ratio == 1:
+        return
+    t = time.perf_counter()
+    for req_id, meta in metadata.reqs_to_save.items():
+        block_ids = meta.local_block_ids
+        for k, v in self.device_kv_caches.items():
+            gb, h, d = v[0].shape
+            indices = torch.tensor(block_ids, device=v[0].device)
+            gbhd = [int(gb / self.block_size), self.block_size, h, d]
+            for i in range(len(self.device_kv_caches[k])):
+                kv = v[i].reshape(gbhd)
+                kv_selected = torch.index_select(kv, 0, indices)
+                bc, bs, h, d = kv_selected.shape
+                shape = int(bs * h / decoder_tp_ratio * d)
+                blocks = torch.chunk(kv_selected, 2, dim=2)
+                vecs = [b.reshape([bc, shape]) for b in blocks]
+                kv_selected = torch.concat(vecs, dim=1).reshape(kv_selected.shape)
+                kv.index_copy_(dim=0, index=indices, source=kv_selected)
```

Comment:
The impl here looks not efficient to me. Does the kv here means host_buffer only?